pkg/subscriptions/insights.jsx

```diff
@@ -0,0 +1,330 @@
+                .catch(catch_error);
+    } else {
+        return cockpit.resolve();
+    }
+}
+
+// TODO - generalize this to arbitrary number of arguments (when needed)
+export function fmt_to_fragments(fmt, arg) {
+    var index = fmt.indexOf("$0");
+    if (index >= 0)
+        return <React.Fragment>{fmt.slice(0, index)}{arg}{fmt.slice(index + 2)}</React.Fragment>;
+    else
+        return fmt;
+}
+
+function left(func) {
+    return function (event) {
+        if (!event || event.button !== 0)
+            return;
+        func();
+        event.stopPropagation();
+    };
+}
+
+export const blurb =
+    _("Proactively identify and remediate threats to security, performance, availability, and stability with Red Hat Insights \u2014 with predictive analytics, avoid problems and unplanned downtime in your Red Hat environment.");
+
+export const link =
+    <a href="https://www.redhat.com/en/technologies/management/insights" target="_blank" rel="noopener">Red Hat Insights <i className="fa fa-external-link" /></a>;
+
+function show_connect_dialog() {
+    show_modal_dialog(
+        {
+            title: _("Connect to Red Hat Insights"),
+            body: (
+                <div className="modal-body">
+                    <strong>{fmt_to_fragments(_("This system is not connected to $0."), link)}</strong>
+                    <p>{blurb}</p>
+                </div>
+            )
+        },
+        {
+            actions: [
+                {
+                    caption: _("Connect"),
+                    style: "primary",
+                    clicked: () => {
+                        return register().catch(catch_error);
+                    }
+                }
+            ]
+        }
+    );
+}
+
+class Revealer extends React.Component {
+    constructor() {
+        super();
+        this.state = { revealed: false };
+    }
+
+    render() {
+        return (
+            <div>
+                <a onClick={event => { if (event.button == 0) this.setState({ revealed: !this.state.revealed }); }}>
+                    {this.props.summary}
+                </a> <i className={this.state.revealed ? "fa fa-angle-down" : "fa fa-angle-right"} />
+                <br />
+                {this.state.revealed && this.props.children}
+            </div>
+        );
+    }
+}
+
+const get_monotonic_start = python.spawn(get_monotonic_start_py).then(data => {
+    return parseFloat(data);
+});
+
+function calc_next_elapse(monotonic_start, timer) {
+    let next_mono = Infinity; let next_real = Infinity;
+    if (monotonic_start) {
+        if (timer.NextElapseUSecMonotonic)
+            next_mono = timer.NextElapseUSecMonotonic / 1e6 + monotonic_start;
+        if (timer.NextElapseUSecRealtime)
+            next_real = timer.NextElapseUSecRealtime / 1e6;
+    }
+    let next = Math.min(next_mono, next_real);
+    if (next !== Infinity)
+        return moment(next * 1000).calendar();
+    else
+        return _("unknown");
+}
```

Comment:
This plus the extra python bit is quite a lot of effort. Why can't this be retrieved from systemd's D-Bus API, like `systemctl list-timers`? That shows the last and next invocations.

Reply:
The numbers are not directly in the API.  I have 'ported' this code over from `systemctl` itself actually.  Look for `calc_next_elapse` in `systemctl.c`.

The Python only looks so bad becasue Python 2 doesn't have `clock_gettime` built in.